s3stream/src/main/java/com/automq/stream/s3/index/SparseRangeIndex.java

```diff
@@ -68,29 +88,70 @@ public void compact(RangeIndex newRangeIndex, Set<Long> compactedObjectIds) {
                 continue;
             }
             if (newRangeIndex != null && !found && rangeIndex.compareTo(newRangeIndex) > 0) {
+                // insert new range index into the list
                 newRangeIndexList.add(newRangeIndex);
                 found = true;
             }
             newRangeIndexList.add(rangeIndex);
         }
         if (newRangeIndex != null && !found) {
+            // insert new range index into the end of the list
             newRangeIndexList.add(newRangeIndex);
         }
-        this.sortedRangeIndexList = newRangeIndexList;
+        int oldSize = size;
+        init(newRangeIndexList);
+        return size - oldSize;
     }
 
-    private void evict() {
-        if (this.sortedRangeIndexList.size() > this.compactNum) {
-            if (evictIndex++ % (sparsePadding + 1) == 0) {
-                this.sortedRangeIndexList.remove(this.sortedRangeIndexList.size() - this.compactNum - 1);
+    /**
+     * Try to evict one range index from the list, the eviction priority for each element is:
+     * 1. any element that's not the first and last N compacted elements
+     * 2. the last N compacted elements
+     * 3. the first element
+     * <p>
+     * For example for a list of [0, 1, 2, 3, 4, 5], compact number is 2, the eviction result will be:
+     * <ul>
+     * <li><code>1rst: [0, 2, 3, 4, 5]</code></li>
+     * <li><code>2nd:  [0, 2, 4, 5]</code></li>
+     * <li><code>3rd:  [0, 4, 5]</code></li>
+     * <li><code>4th:  [0, 5]</code></li>
+     * <li><code>5th:  [0]</code></li>
+     * <li><code>6th:  []</code></li>
+     * </ul>
+     *
+     * @return evicted size
+     */
+    public int evictOnce() {
+        int indexToEvict = -1;
+        if (this.sortedRangeIndexList.isEmpty()) {
+            return 0;
+        } else if (this.sortedRangeIndexList.size() == 1) {
+            // evict the only element
+            indexToEvict = 0;
+        } else if (this.sortedRangeIndexList.size() <= (1 + compactNum)) {
+            indexToEvict = 1;
+        }
+
+        if (indexToEvict == -1) {
```

Comment:
If the evictIndex = 0, sortedRangeIndexList.size() == 2 and compactNum = 2, then the sortedRangeIndexList.remove will throw IndexOutOfBoundException